Support change github webapp

bzhaoopenstack opened the pull request:
This patch uses the original way to change github webapp by python crawler.
So we have to provider the github user password, as we can not use the
oauth token to simulate the human user login, also the github api doesnt
provide a way to change the app webhook. So python crawler is the only
to implement this.

1. Add 3 more option into local config, including github username, user
password, and app name.
2. Also add test and normal case for labkeeper to test, such as moo-ai

Related-Bug: theopenlab/openlab#218

Commit:
Support change github webapp

This patch uses the original way to change github webapp by python crawler.
So we have to provider the github user password, as we can not use the
oauth token to simulate the human user login, also the github api doesnt
provide a way to change the app webhook. So python crawler is the only
to implement this.

1. Add 3 more option into local config, including github username, user
password, and app name.
2. Also add test and normal case for labkeeper to test, such as moo-ai

Related-Bug: theopenlab/openlab#218

huangtianhua commented:
recheck

bzhaoopenstack commented:
recheck

bzhaoopenstack commented:
recheck

wangxiyuan commented on etc/ha_healthchecker/ha_healthchecker.py.j2:
```diff
@@ -25,9 +27,15 @@ CONFIG_LOCAL = {
     'github_user_token': "{{ test_github_token }}",
+    'github_user_passwd': "{{ test_github_password }}",
+    'github_user_name': "{{ test_github_user_name }}",
+    'github_app_name': "{{ test_github_app_name }}"
```
please add them into openlabcmd

Commit:
Use base64 in zk configuration

We will encode 3 config options:
'github_user_passwd'
'dns_provider_token'
'github_user_token'

wangxiyuan commented on etc/ha_healthchecker/ha_healthchecker.py.j2:
```diff
@@ -25,9 +28,15 @@ CONFIG_LOCAL = {
     'github_user_token': "{{ test_github_token }}",
+    'github_user_passwd': "{{ test_github_password }}",
+    'github_user_name': "{{ test_github_user_name }}",
+    'github_app_name': "{{ test_github_app_name }}"
```
what's the value of test_github_app_name? we may create one in moo?

bzhaoopenstack replied:
oh, I forget to add this in the inventory.

bzhaoopenstack replied:
how about named "test-ha-app"?

wangxiyuan replied:
++

wangxiyuan commented on etc/ha_healthchecker/ha_healthchecker.py.j2:
```diff
@@ -780,23 +861,59 @@ class Switcher(Base):
                     value = CONFIG_LOCAL.get(attr)
+                else:
+                    if attr in ['github_user_passwd', 'dns_provider_token',
+                                'github_user_token']:
```
it's good set a variable for the list, like `base64_encoded_options` so that we can update it easily in the future just in one place.

bzhaoopenstack replied:
done

huangtianhua commented on etc/ha_healthchecker/ha_healthchecker.py.j2:
```diff
@@ -25,9 +28,15 @@ CONFIG_LOCAL = {
     'github_repo': 'theopenlab/openlab',
     'github_user_token': "{{ github_token }}",
+    'github_user_passwd': "{{ github_password }}",
+    'github_user_name': "theopenlab-ci",
```
1. Why not use {{ github_username}} here instead of hard code?
2. I don't think we have the var 'github_password' anymore, we use github_token instead of password, right?

bzhaoopenstack replied:
Yeah, I'm very clear what is done here, and I know very clear not use the password here.  But we could get only 1 way to change app webhook. Github API doesn't provide an API to change it[1].. So we only can simulate a human user to login github and change the app webhook by a python crawlers

reference:
[1] https://developer.github.com/v3/apps/

huangtianhua commented on etc/ha_healthchecker/ha_healthchecker.py.j2:
```diff
@@ -25,9 +28,15 @@ CONFIG_LOCAL = {
 {% if use_test_account %}
     'github_repo': "{{ test_repo_name }}",
     'github_user_token': "{{ test_github_token }}",
+    'github_user_passwd': "{{ test_github_password }}",
```
So here will be better to use github auth token instead?

bzhaoopenstack replied:
There is no way to use oauth token to update the github app webhook, github api doesnt provider that. So we use python crawlers to simulate the normal human to login Github and change the webhook on a browser..

Commit:
Fix COMMENTS

Add new var "github_appname" in all.yaml

wangxiyuan commented:
Get it merge since maybe nobady knows what's is done here. lol
